usr/src/cmd/zdb/zdb.c

```diff
@@ -2121,6 +2130,153 @@ dump_label_uberblocks(vdev_label_t *lbl, uint64_t ashift)
 	}
 }
 
+static int
+dump_specific_file(objset_t *os, uint64_t obj)
+{
+	int print_header = 0;
+
+	if (dump_opt['v'] > 5)
+		print_header = 1;
+	dump_object(os, obj, dump_opt['v'], &print_header);
+	return (0);
+}
+
+static int
+dump_specific_directory(objset_t *os, uint64_t obj, char *name)
+{
+	int err;
+	uint64_t child_obj;
+	char *slash = strchr(name, '/');
+	char *namep = name;
+	dmu_buf_t *db;
+	dmu_object_info_t doi;
+
+	if (slash != NULL)
+		*slash = '\0';
+	err = zap_lookup(os, obj, namep, 8, 1, &child_obj);
+	if (slash != NULL)
+		*slash = '/';
+
+	if (err != 0) {
+		fprintf(stderr, "Looking up '%s' failed: %s\n",
+		    name, strerror(err));
+		return (err);
+	}
+
+	child_obj = ZFS_DIRENT_OBJ(child_obj);
+	err = sa_buf_hold(os, child_obj, FTAG, &db);
+	if (err != 0) {
+		fprintf(stderr, "Can't get SA dbuf for obj %lu: %s\n",
+		    obj, strerror(err));
+		return (EINVAL);
+	}
+	dmu_object_info_from_db(db, &doi);
+	sa_buf_rele(db, FTAG);
+
+	if (dump_opt['v'] > 6)
+		fprintf(stderr, "%s: obj %lu name '%s' type %d bonustype %d\n",
+		    __func__, obj, name, doi.doi_type, doi.doi_bonus_type);
+
+	if (doi.doi_bonus_type != DMU_OT_SA &&
+	    doi.doi_bonus_type != DMU_OT_ZNODE) {
+		fprintf(stderr, "Invalid bonus type %d for obj %lu!\n",
+		    doi.doi_bonus_type, obj);
+		return (EINVAL);
+	}
+
+	switch (doi.doi_type) {
+	case DMU_OT_PLAIN_FILE_CONTENTS:
+		ASSERT3P(slash, ==, NULL);
```

Comment:
Not a very nice way to communicate a user error (imagine `zdb -o file pool/fs /dir/file/impossible"`, but I guess this is just zdb :-\